migration/migrator.go

```diff
@@ -0,0 +1,347 @@
+	mpt, err := trie.NewStateTrie(id, m.mptdb)
+	if err != nil {
+		return common.Hash{}, err
+	}
+
+	zkt, err := trie.NewZkMerkleStateTrie(zkStorageRoot, m.zkdb)
+	if err != nil {
+		return common.Hash{}, err
+	}
+
+	var mu sync.Mutex
+	var slots atomic.Uint64
+	err = hashRangeIterator(zkt, NumProcessStorage, func(key, value []byte) error {
+		slots.Add(1)
+		slot := m.readZkPreimage(key)
+		trimmed := common.TrimLeftZeroes(common.BytesToHash(value).Bytes())
+		mu.Lock()
+		defer mu.Unlock()
+		if err := mpt.UpdateStorage(address, slot, trimmed); err != nil {
+			return err
+		}
+		log.Trace("Updated storage slot to MPT", "contract", address.Hex(), "index", common.BytesToHash(key).Hex())
+		return nil
+	})
+	if err != nil {
+		return common.Hash{}, err
+	}
+
+	root, err := m.commit(mpt, types.EmptyRootHash)
+	if err != nil {
+		return common.Hash{}, err
+	}
+	log.Debug("Storage migration finished", "account", address, "slots", slots.Load(), "elapsed", time.Since(startAt))
+	return root, nil
+}
+
+func (m *StateMigrator) readZkPreimage(key []byte) []byte {
+	hk := *trie.IteratorKeyToHash(key, true)
+	if preimage, ok := m.allocPreimage[hk]; ok {
+		return preimage
+	}
+	if preimage := m.zkdb.Preimage(hk); preimage != nil {
+		if common.BytesToHash(zk.MustNewSecureHash(preimage).Bytes()).Hex() == hk.Hex() {
+			return preimage
+		}
+	}
+	panic("preimage does not exist: " + hk.Hex())
+}
+
+func (m *StateMigrator) commit(mpt *trie.StateTrie, parentHash common.Hash) (common.Hash, error) {
+	root, set, err := mpt.Commit(true)
+	if err != nil {
+		return common.Hash{}, err
+	}
+	if set == nil {
+		log.Warn("Tried to commit state changes, but nothing has changed.", "root", root)
+		return root, nil
+	}
+
+	// NOTE(pangssu): It is possible that the keccak256 and poseidon hashes collide, and data loss can occur.
+	for path, mptNode := range set.Nodes {
+		data, _ := m.db.Get(mptNode.Hash.Bytes())
+		if len(data) == 0 {
+			continue
+		}
+		if node, err := zk.NewTreeNodeFromBlob(data); err == nil {
+			return common.Hash{}, fmt.Errorf("Hash collision detected: hashKey: %v, path: %v, data: %v, zkNode: %v", mptNode.Hash, path, data, node)
+		}
+	}
+
+	if err := m.mptdb.Update(root, parentHash, 0, trienode.NewWithNodeSet(set), nil); err != nil {
+		return common.Hash{}, err
+	}
+	if err := m.mptdb.Commit(root, false); err != nil {
+		return common.Hash{}, err
+	}
+	return root, nil
+}
+
+func (m *StateMigrator) FinalizeTransition(transitionBlock types.Block) {
+	// We need to update the chain config to set the correct hardforks.
+	genesisHash := rawdb.ReadCanonicalHash(m.db, 0)
+	cfg := rawdb.ReadChainConfig(m.db, genesisHash)
+	if cfg == nil {
+		panic("chain config not found")
+	}
+
+	// Set the standard options.
+	cfg.LondonBlock = transitionBlock.Number()
+	cfg.ArrowGlacierBlock = transitionBlock.Number()
+	cfg.GrayGlacierBlock = transitionBlock.Number()
+	cfg.MergeNetsplitBlock = transitionBlock.Number()
```

Comment:
Curious what these codes do, could you explain to me?

Reply:
Part of this code comes from the OP's bedrock migration code, they set values for some hardforks. 
But it doesn't seem necessary for us, so i'll delete it.

https://github.com/ethereum-optimism/optimism-legacy/blob/develop/op-chain-ops/genesis/db_migration.go#L285